mycpp/gc_mylib.cc

```diff
@@ -129,66 +140,42 @@ bool CFileWriter::isatty() {
 }
 
 //
-// Buf
+// BufWriter
 //
 
-// TODO: Consider renaming MutableStr, or make this a subclass of Str
-class Buf : Obj {
- public:
-  // The initial capacity is big enough for a line
-  Buf(int cap) : Obj(Tag::Opaque, kZeroMask, 0), len_(0), cap_(cap) {
-  }
-  void Extend(Str* s);
-
- private:
-  friend class BufWriter;
-  friend Str* StrFromBuf(const Buf*);
-  friend Buf* NewBuf(int);
-
-  // TODO: move this state into BufWriter
-  int len_;  // data length, not including NUL
-  int cap_;  // capacity, not including NUL
-  char data_[1];
-};
-
-Str* StrFromBuf(const Buf* buf) {
-  return ::StrFromC(buf->data_, buf->len_);
+char* BufWriter::data() {
+  assert(str_);
+  return str_->data_;
 }
 
-Buf* NewBuf(int cap) {
-  // TODO: sizeof(Buf) is an overestimate because of flexible array member
-  void* place = gHeap.Allocate(sizeof(Buf) + cap + 1);
+char* BufWriter::end() {
+  assert(str_);
+  return str_->data_ + len_;
+}
 
-  auto* b = new (place) Buf(cap);
-  return b;
+int BufWriter::capacity() {
+  return str_ ? len(str_) : 0;
 }
 
-void Buf::Extend(Str* s) {
+void BufWriter::Extend(Str* s) {
```

Comment:
I would be inclined to inline Extend() into write(), since it's only used in one place and is very short, but that is just a suggestion

Reply:
Done